extensions/amp-content-viewer/0.1/amp-content-viewer.js

```diff
@@ -0,0 +1,696 @@
+      this.gestures_.cleanup();
+      this.gestures_ = null;
+    }
+  }
+
+  /** @private */
+  setupGestures_() {
+    // TODO (#12881): this and the subsequent use of event.preventDefault
+    // is a temporary solution to #12362. We should revisit this problem after
+    // resolving #12881 or change the use of window.event to the specific event
+    // triggering the gesture.
+    this.gestures_ = Gestures.get(
+        this.element,
+        /* opt_shouldNotPreventDefault */ false
+    );
+
+    this.gestures_.onPointerDown(() => {
+      if (this.motion_) {
+        this.motion_.halt();
+      }
+    });
+
+    // Zoomable.
+    this.gestures_.onGesture(DoubletapRecognizer, e => {
+      const newScale = this.scale_ == 1 ? this.maxScale_ : this.minScale_;
+      const deltaX = this.elementBox_.width / 2 - e.data.clientX;
+      const deltaY = this.elementBox_.height / 2 - e.data.clientY;
+      this.onZoom_(newScale, deltaX, deltaY, true).then(() => {
+        return this.onZoomRelease_();
+      });
+    });
+
+    this.gestures_.onGesture(TapzoomRecognizer, e => {
+      this.onTapZoom_(e.data.centerClientX, e.data.centerClientY,
+          e.data.deltaX, e.data.deltaY);
+      if (e.data.last) {
+        this.onTapZoomRelease_(e.data.centerClientX, e.data.centerClientY,
+            e.data.deltaX, e.data.deltaY, e.data.velocityY, e.data.velocityY);
+      }
+    });
+
+    this.gestures_.onGesture(PinchRecognizer, e => {
+      this.onPinchZoom_(e.data.centerClientX, e.data.centerClientY,
+          e.data.deltaX, e.data.deltaY, e.data.dir);
+      if (e.data.last) {
+        this.onZoomRelease_();
+      }
+    });
+
+    // Override all taps
+    this.gestures_.onGesture(TapRecognizer, e => {
+      const event = createCustomEvent(this.win, 'click', null, {bubbles: true});
+      e.data.target.dispatchEvent(event);
+    });
+  }
+
+  /**
+   * Registers a Swipe gesture to handle panning when the content is zoomed.
+   * @private
+   */
+  registerPanningGesture_() {
+    // Movable.
+    this.unlistenOnSwipePan_ = this.gestures_
+        .onGesture(SwipeXYRecognizer, e => {
+          event.preventDefault();
+          this.onMove_(e.data.deltaX, e.data.deltaY, false);
+          if (e.data.last) {
+            this.onMoveRelease_(e.data.velocityX, e.data.velocityY);
+          }
+        });
+  }
+
+  /**
+   * Deregisters the Swipe gesture for panning when the content is zoomed out.
+   * @private
+   */
+  unregisterPanningGesture_() {
+    if (this.unlistenOnSwipePan_) {
+      this.unlistenOnSwipePan_();
+      this.unlistenOnSwipePan_ = null;
+      this.gestures_.removeGesture(SwipeXYRecognizer);
+    }
+  }
+
+  /**
+   * Returns value bound to min and max values +/- extent.
+   * @param {number} v
+   * @param {number} min
+   * @param {number} max
+   * @param {number} extent
+   * @return {number}
+   * @private
+   */
+  boundValue_(v, min, max, extent) {
+    return Math.max(min - extent, Math.min(max + extent, v));
```

Comment:
Is this different than `math.js#clamp(v, min - extent, max + extent)`?

Reply:
nit: what is `v`? is it `value`?

Reply:
Yup, this is basically a clamp and we can import that. `v` is bad name for `value`. Fixed.